### PR TITLE
Adding logging

### PR DESCRIPTION
**Pull request by mmontgomery-square**

### CODE REVIEW

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli.go:40]: In addition to the marquee changes, I went through and updated all of the errors so that they're properly messaged and wrapped (wrapping gives us stack traces).

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli.go:87]: I wanted a single choke point to ensure logging always does the right thing, and since the original Run method had a couple exit points I thought wrapping it made sense. However... given that Go has defer, perhaps it would be better to put this logic in a defer block?

File: cmd/spiffe-aws-assume-role/cli/cli_test.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli_test.go:36]: After ensuring that we start from a clean slate, we pass in a purposely bad command line argument and then verify that it shows up in the logs.

**Comment by captiosus** [comment]
Could you attach the logger to the connection to the local spire-agent as well: https://github.com/square/spiffe-aws-assume-role/blob/master/pkg/credentials/spiffejwt.go#L40

Would be helpful for debugging connection issues there

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mbyczkowski) [inline cmd/spiffe-aws-assume-role/cli/cli.go:113]: I wonder if we should try to log to `os.Stderr` by default (to make sure only the AWS credentials go to `os.Stdout`) and make file output optional?

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [reply cmd/spiffe-aws-assume-role/cli/cli.go:39]: The "Wrap" part is misleading, what I'm really doing is capturing the stack trace so that it's available for debugging later. When I was initially adding tests to the package the native error values would contain messages but not the stack traces, which made debugging difficult. errors.Wrap fixes that and makes debugging super easy.

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [reply cmd/spiffe-aws-assume-role/cli/cli.go:113]: Per conversation with Roy, since this is all executed via scripting there needed to be a durable place for us to put errors so a human could inspect it after the fact. However, I do think it makes sense to try to log to both a file and standard error.

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli.go:36]: Since we need to provide a Logger instance to the SPIRE agent connection and not just on application exit, we're promoting the Logger instance to be instantiated up-front and long-lived by sticking it in the CliContext.

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli.go:95]: Really nice improvement over the previous revisions: combining named return values and defer to eliminate the need for a wrapper function to handle error logging.

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli.go:119]: Two things here:

First, the code that I had copied from the Logrus website was wrong, if you look at the source of the New() function (do we still call them constructors?) you can see that it actually assigns os.Stderr (not stdout) by default so not only was it the wrong stream but we also didn't actually need to do any overwriting.

Second, and more importantly, we're now using a MultiWriter (TIL) to fork output to both stderr and the log file.

File: pkg/credentials/jwt_source_provider.go
Comment (mmontgomery-square) [inline pkg/credentials/jwt_source_provider.go:27]: Adding a Logger as a parameter. Note that this is the SPIFFE Logger and not the Logrus Logger, although Logrus is compatible with the SPIFFE Logger's interface. If we want to push the SPIFFE logger down farther and not expose it via the signature, I wouldn't be against it.

File: pkg/credentials/spiffejwt.go
Comment (mmontgomery-square) [inline pkg/credentials/spiffejwt.go:50]: Fortunately adding logging is just a one-liner.

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (captiosus) [inline cmd/spiffe-aws-assume-role/cli/cli.go:19]: Can we make this configurable through the CLI so we internally we can choose where to write the logs. Also, this might be better as a file path

Comment (mmontgomery-square) [reply cmd/spiffe-aws-assume-role/cli/cli.go:19]: Will do

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli.go:74]: Our new behavior:

1. Create a default Logger instance that emits text-formatted logs to stderr
2. If the --log-file-path parameter is specified, also emit logs to that file

We accomplish this by updating the configuration of the passed-in Logger after parsing is complete.

File: cmd/spiffe-aws-assume-role/cli/cli.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli.go:84]: I thought it might be helpful to have an example of what the logs look like.

File: cmd/spiffe-aws-assume-role/cli/cli_test.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli_test.go:28]: In an effort to reduce boilerplate (many of the tests just take the standard required args and add one more parameter) I've extracted the required arguments to their own variable so that they don't have to be repeated over and over.

File: cmd/spiffe-aws-assume-role/cli/cli_test.go
Comment (mmontgomery-square) [inline cmd/spiffe-aws-assume-role/cli/cli_test.go:69]: As a result of boilerplate reduction/introduction of helper methods, basic parsing tests are much more concise.

**Review by captiosus — APPROVED** [review]
LGTM after fixing the merge conflict

**Pull request merged**
